src/Functions/FunctionsJSON.h

```diff
@@ -984,13 +985,19 @@ class JSONExtractImpl
             throw Exception{"Function " + String(function_name) + " requires at least two arguments", ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH};
 
         const auto & col = arguments.back();
-        auto col_type_const = typeid_cast<const ColumnConst *>(col.column.get());
-        if (!col_type_const || !isString(col.type))
-            throw Exception{"The last argument of function " + String(function_name)
-                                + " should be a constant string specifying the return data type, illegal value: " + col.name,
-                            ErrorCodes::ILLEGAL_COLUMN};
+        if (isString(col.type))
+        {
+            auto col_type_const = typeid_cast<const ColumnConst *>(col.column.get());
+            if (col_type_const)
+                return DataTypeFactory::instance().get(col_type_const->getValue<String>());
+
+            if (col.column && (col.column->size() == 1))
```

Comment:
So it relies on the behavior of default implementation of nullable arguments.  Worth adding some comments.

Reply:
I think I'll change my solution because the query `SELECT JSONExtract(toNullable('{"string_value":null}'), 'string_value', 'String')` returns an empty string of type `Nullable(String)` instead of `NULL` which is more expected for this query.

Reply:
It's getting complicated. Let's get back to your solution.